### ISSUE TITLE
Discussion: Have you seen real projects using WinUI?

### ISSUE THREAD

**Issue by omikhailov** [microsoft/microsoft-ui-xaml#6791]
I heard that Windows 11 preinstalled apps are made with WinUI 2, and they are being developed by a team from the Chinese office, but it would be interesting to know how and where folk use WinUI elsewhere. Do you know about commercial projects built with WinUI? What kind of apps were they? Was it a good experience?

P.S. Sorry for the flooding, I'm just curious.

**Comment by shelllet** [comment]
I have developed a simple `rpa` tool by c++/winui3, but I would like to say that the current progress and featurs of winui3 cannot develop commercial applications,
![image](https://user-images.githubusercontent.com/16729017/156960353-503bdd74-967c-4dd4-9408-d8f2285ba7f5.png)

**Comment by llothar** [comment]
I'm listening to a lot of solo developers in the independent app scene (former known as shareware). 
And i can answer with a clear no.  At the moment nobody i know is even planning to go this way.

**Comment by dongle-the-gadget** [comment]
Well there are many production apps currently using WinUI 2, not sure about WinUI 3 though.

**Comment by levinli303** [comment]
I have an unarchiver app written in WinUI3, also available on Microsoft Store. It's cross platform, sharing some objective-c(++) code under the hood with iOS/macOS. The UI part is entirely WinUI3. It is lacking some features for sure (drag and drop virtual file without needing a dll extension), but mostly it is ok as an unarchiver.

https://www.microsoft.com/en-us/p/easy-unarchiver/9nkl1jq5c62v

**Comment by MEK3DK** [comment]
As others here say WinUI 3 is not ready for commercial applications. 

We did an investigation into WinUI 3, whether we could start changing some apps to use WinUI 3 instead of WPF, but there are a lot of features missing and the amount of bugs makes it unrealistic. 

WinUI 3 is only for hobby apps atm, my best estimate is 2 years untill it might be ready, depending on what resources Microsoft apply.

**Comment by fredemmott** [comment]
Hobbyist with a cross-platform (mostly Qt, little bit of wxWidgets) background; I just ported my app to WinUI3 with C++/WinRT.

There's a lot of rough edges in winui, winui's documentation, and MSIX, especially when not distributing in the MS store. MSIX is not part of WinUI, but there's such a bias towards 'packaged' apps in WinUI and the App SDK that it feels fair to bundle.

Overall, Qt is leagues ahead on documentation and raw power/usability, but I'm avoiding it given I'm not confident in their current stance on open source. Compared to wx, winui makes the easy stuff hard and the hard stuff easy. It's much easier for me to make something work with wx, but once I have something working, much easier to make it look good and feel usable with winui (except for title bars).

Concretely:

* navigation view + navbar + title bar content/back button doesn't work without a lot of manual work. I suggest MS attempt to dogfood this in the winui 3 controls gallery  - which - at least in the store version - has a classic windows title bar on Windows 10.
* if you start a winui3 app (including the controls gallery) and don't move the mouse, it will show a busy spinning cursor until you move the mouse. This seems trivial, but makes winui3 apps 'feel' slow
* many things are technically documented, but hard to find -  e.g. how to disable the mddbootstrapinitializer to do it by hand, how to disable the xaml generated main. Most of this is documented on /some/ page, but it's quicker for me to dig into the source than attempt to find the documentation. Documentation is spread between c++/winrt, windows app sdk, and winui 3, which makes sense - but it would be much easier if documentation were sorted by high level problem rather than the component that solves the problem. API reference documentation is great once you know what class/function/component/... you need, but finding that out (discoverability) is the problem.
* many times after searching for documentation on MSDN, I simply don't find it, so I google. Many small-but-key things (e.g. that `IWindowNative` exists, or that with winui contentdialogs from c++/winrt you need XamlRoot + `IInitializeWithWindow`) have reference documentation, but it's very hard to find on MSDN that they exist, and that they solve core problems. Google searches almost always end up pointing me to an Old New Thing blog post or a StackOverflow answer by Raymond Chen - not the official documentation. While Raymond Chen's blog and answers are incredibly valuable, they should not be an essential resource for basic usage.
* the example code in the winui controls gallery does not always line up with the actual code, and does not actually work. 
* some things are completely undocumented, e.g. the fact that a NumberBox NumberFormatter must be an `INumberParser`, not just an `INumberFormatter2`
* errors are uninformative, e.g. if you attempt to set NumberBox::NumberFormatter to an `INumberFormatter2` which isn't an `INumberParser`, you just get 'failed to set property'
* it's hard to find the link between the style guidelines (typography, colors etc) and "for these controls, these styles exist in the theme resource". As far as I can tell, the official reference is https://github.com/microsoft/microsoft-ui-xaml/tree/master/dev/CommonStyles - if you're not already familiar with xaml and styles, it takes a while to understand how these relate to what you'll see.

Ecosystem things:
* the build-time distinction between 'packaged' and 'unpackaged' seems much more pervasive than it needs to be:
  * If I want to support .zip or .msix installation, AIUI I 'should' have two `.vcxproj` things, and #ifdefs in my code, and keep track of twice as many .pdb files to debug user crashes.
  * Once you dig through the source code and find out what flags need setting, it's relatively straightforward to make a single winui3 exe that works both packaged and unpackaged: https://github.com/microsoft/WindowsAppSDK/issues/2216
* packaged is 'recommended', but if you're not using the MS store, key logistical features are currently only available on windows 11 insider builds, and it's unclear if they will ever be available on windows 10, e.g.:
  * support for automatic updates: https://docs.microsoft.com/en-us/windows/msix/app-installer/auto-update-and-repair--overview
  * being able to replace code signing keys when they expire: https://docs.microsoft.com/en-us/windows/msix/package/persistent-identity
* If you're a cross-platform developer that happens to be working on a windows-specific app (or a cross-platform app with platform-specific frontends), the coupling to visual studio/nuget/msbuild is an extra painful step; c++/winrt is great and feels like an attempt to engage with and attract the broader C++ community, but the windows app sdk + winui coupling to MS-specific tools feels contradictory and like a lock-in attempt. MSBuild in particular is unavoidable, due to it being the only supported way to invoke the xaml compiler. For C++, CMake is the standard.
  
Despite all of that: I'm happy overall with the results, and would *probably* do it again for a hobby app, but I definitely wouldn't consider it for a larger project - or any project with multiple developers - at the moment.

![image](https://user-images.githubusercontent.com/360927/157482107-e269f646-25a6-4fb5-97d7-f71b0b9455f6.png)

**Comment by StephenLPeters** [comment]
@ryandemopoulos FYI

**Comment by wherewhere** [comment]
Mine! Mine! Mine! [APK Installer](https://github.com/Paving-Base/APK-Installer)
![mmexport1645260410258.png](https://user-images.githubusercontent.com/27689196/157800968-3272a2fc-28f8-4f79-b4a5-4450abc28ee5.png)
However, I use it only because it can use process. In fact, I don't like Windows App SDK at all. It write like UWP, looks like UWP. But it doesn't feel like UWP. What we need is a UWP which has the same permission and functions as WPF. Not a WPF which only looks like UWP. I want UWP back 😭.

**Comment by ryandemopoulos** [comment]
This is good timing, so thanks for starting the discussion @omikhailov -- our community call next week will talk to this a bit. We have a couple partnerships with substantial apps (Fortune 500 companies that aren't Microsoft) happening right now, but I can't share specifics due to agreements with those companies. One or perhaps both of those apps will ship this year and we'll be able to talk about them once they are public.

As mentioned, most of the inbox apps in Windows use WinUI 2 today -- one of them is very close to being converted to WinUI 3 already, and others will move to WinUI 3 over time. Tune in to the Wednesday call next week as we'll be releasing a new roadmap for the WinAppSDK and talking more about how we're prioritizing getting WinAppSDK and WinUI 3 into more industry-leading apps this year.

**Comment by wbokkers** [comment]
Xelion is using WinUI 3 for their Windows app. You need a Xelion account to run it.
https://appinstaller.xelion.com/winapp/v8/

**Comment by torquerrr** [comment]
@ryandemopoulos 
> one of them is very close to being converted to WinUI 3 already, and others will move to WinUI 3 over time
Do you mean that internal Windows apps that are currently "Win32/UWP + WinUI 2" (e.g. Settings, Store, the shell) will be converted to "Win32/WPF/etc. + WinUI 3" over time?
Thanks.

**Comment by JanRajnoha** [comment]
I'm currently creating cash system in WinUI 3 and sometimes it is really hard to solve problems :/ on the other hand, every time i found solution

**Comment by charlesroddie** [comment]
> key logistical features are currently only available on windows 11 insider builds, and it's unclear if they will ever be available on windows 10, e.g.:
> support for automatic updates: https://docs.microsoft.com/en-us/windows/msix/app-installer/auto-update-and-repair--overview

Support for automatic updates has always been the key advantage of MSIX. I think this may be a misunderstanding of the confusingly worded documentation, and that the doc banner ("The following article discusses settings currently available in Windows Insider build 22415 and newer") really means that "this page applies to Windows Insider build 22415: some of these features work on Win10 but we are not telling you which ones!".

**Comment by octaviordz** [comment]
Since documentation is mentioned in this thread, has that improved? I'm trying to find WinUI3 documentation without much success. I see the documentation on the API but, is there any high level documentation for example on how the layout system work, what is the designed way to compose the controls?

Thanks.

**Comment by wbokkers** [comment]
In App Center I can see that only 0.1 to 0.15% of the users or our Xelion app (about 30k users) see crashes. 
Currently, we don't get many support tickets about crashes. 

So, WinUI 3 can be successfully used in a real world app. But there is still room for improvement.

**Comment by fredemmott** [comment]
> So, WinUI 3 can be successfully used in a real world app

Yeah, it depends on what features you use, and what your app does; but there isn't a way to find out if your app design is possible to do in winui 3 without crashing except by trying it and seeing if you're lucky.

**Comment by wbokkers** [comment]
I can almost safely assume that our app is among the most complex WinUI 3 apps available. We use almost every corner of the WinUI3 library. It's not a simple forms over data app.

**Comment by tajbender** [comment]
> I’m still using winui 3 for my hobby project, but this is only because as a hobby project I don’t currently have the time to rewrite the UI layer.

Well, coming from WinForms I never used XAML before, so it was a hard time for me.. 

However, with some patience I've learned some patterns and I love the layout system, as I love _Hot Deploy_ and _Hot Reload_. So sure, it's a lot of try & error.

But there are Open Books, published by Microsoft, and - no joke - https://www.bing.com/search?q=chatgpt can help, too! 🚀

**Comment by fredemmott** [comment]
> I can almost safely assume that our app is among the most complex WinUI 3 apps available. We use almost every corner of the WinUI3 library. It's not a simple forms over data app.

The facts are that there are known crashes in WinUI3, in the issue tracker here, acknowledged by Microsoft, and not showing signs of being worked on. I'm glad your app doesn't happen to trigger them.

My meaning is not "only simple apps can be done in WinUI3" - it's "doing a complex app in WinUI3 is a risk - you might get lucky, or you might be in an unfixable hole".

**Comment by wbokkers** [comment]
I agree there's lot of rough edges and much room for improvement. But I am not just feeling lucky I didn't ran into an issue. I ran into many issues... and fixed them one way or another. This is far from ideal. WinUI 3 should be stable in all circumstances and it isn't.

**Comment by llothar** [comment]
Well the sad fact is that the good old rule that a new software technology still needs 5 years to be useable and 10 years to mature are real. WinUI3 is in year 1.5 now.  So we need patient to see it become bug free.  

By the way, Swift-UI is almost as unuseable as WinUI3 and both companies have almost unlimited money they could throw at the projects. But they know it wont help to speed things up.

**Comment by tajbender** [comment]
time passes, updates pass

**Comment by legistek** [comment]
First off, WinUI is based off of tech that was first released in 2012 - WinRT - so the "it takes time" excuse doesn't work. They had to reboot it twice because it was a commercial and technological failure. WinUI3 is in fact the third flavor of WinRT but so far appears to be faring no better than its predecessors. Someone at the company is obviously determined to keep the tech alive at all costs when all evidence is that it should've been scrapped 10 years ago.

That they HAD an amazing, mature, usable framework for commercial applications - WPF - but rather than modernize it they put it out to pasture for something newer and shinier. 

Meanwhile what does Microsoft use for its own commercial Windows applications? It sure isn't anything that's available to us, which is I'm sure how their application teams like it.
